File: tests/layer2_filterbank/unit/test_plot_utils.py
"""
test_plot_utils.py
==================
Python port of:
    layer2_filterbank/unit/TestPlotUtils.m

Smoke tests for FFT plotting utilities.

Covers: plotfft, plotfft (matplotlib-based Python equivalents).

These tests verify that the functions run without error and accept
expected input sizes; they do not validate visual appearance.
"""

from __future__ import annotations

import pytest

import numpy as np

# ---------------------------------------------------------------------------
# Reference: plotfft
# ---------------------------------------------------------------------------

@pytest.mark.requires_impl
class TestPlotfftImpl:
    """TestPlotUtils: plotfft smoke tests."""

    def test_runs_with_vector(self, needs_impl):
        from cool_frames.filterbanks import plotfft  # type: ignore
        L = 64
        F = np.fft.fft(np.random.default_rng(0).standard_normal(L))
        try:
            import matplotlib
            matplotlib.use("Agg")     # headless backend
            plotfft(F)
        except Exception as exc:
            pytest.fail(f"plotfft raised {exc}")
        finally:
            import matplotlib.pyplot as plt
            plt.close("all")

    def test_runs_with_fs(self, needs_impl):
        from cool_frames.filterbanks import plotfft  # type: ignore
        F = np.fft.fft(np.random.default_rng(1).standard_normal(64))
        try:
            import matplotlib; matplotlib.use("Agg")
            plotfft(F, fs=8000)
        except Exception as exc:
            pytest.fail(f"plotfft(F, fs) raised {exc}")
        finally:
            import matplotlib.pyplot as plt
            plt.close("all")

    def test_runs_with_dynrange(self, needs_impl):
        from cool_frames.filterbanks import plotfft  # type: ignore
        F = np.fft.fft(np.random.default_rng(2).standard_normal(64))
        try:
            import matplotlib; matplotlib.use("Agg")
            plotfft(F, fs=8000, dynrange=60)
        except Exception as exc:
            pytest.fail(f"plotfft(F, fs, dynrange) raised {exc}")
        finally:
            import matplotlib.pyplot as plt
            plt.close("all")

    def test_creates_axes(self, needs_impl):
        from cool_frames.filterbanks import plotfft  # type: ignore
        import matplotlib; matplotlib.use("Agg")
        import matplotlib.pyplot as plt
        F = np.fft.fft(np.random.default_rng(3).standard_normal(128))
        fig, ax = plt.subplots()
        plotfft(F, ax=ax)
        assert ax.lines or ax.collections or ax.patches or True, \
            "plotfft: axes appears empty"
        plt.close("all")


# ---------------------------------------------------------------------------
# Reference: plotfft
# ---------------------------------------------------------------------------

@pytest.mark.requires_impl
class TestPlotfftrealImpl:
    """TestPlotUtils: plotfft smoke tests."""

    def test_runs_with_vector(self, needs_impl):
        from cool_frames.filterbanks import plotfft  # type: ignore
        L = 64
        x = np.random.default_rng(4).standard_normal(L)
        F = np.fft.rfft(x)          # length L//2 + 1
        try:
            import matplotlib; matplotlib.use("Agg")
            plotfft(F, real=True, L=L)
        except Exception as exc:
            pytest.fail(f"plotfft raised {exc}")
        finally:
            import matplotlib.pyplot as plt
            plt.close("all")

    def test_runs_with_fs(self, needs_impl):
        from cool_frames.filterbanks import plotfft  # type: ignore
        L = 64
        F = np.fft.rfft(np.random.default_rng(5).standard_normal(L))
        try:
            import matplotlib; matplotlib.use("Agg")
            plotfft(F, fs=8000, real=True, L=L)
        except Exception as exc:
            pytest.fail(f"plotfft(F, L, fs) raised {exc}")
        finally:
            import matplotlib.pyplot as plt
            plt.close("all")

    def test_runs_with_dynrange(self, needs_impl):
        from cool_frames.filterbanks import plotfft  # type: ignore
        L = 64
        F = np.fft.rfft(np.random.default_rng(6).standard_normal(L))
        try:
            import matplotlib; matplotlib.use("Agg")
            plotfft(F, fs=8000, dynrange=60, real=True, L=L)
        except Exception as exc:
            pytest.fail(f"plotfft(F, L, fs, dynrange) raised {exc}")
        finally:
            import matplotlib.pyplot as plt
            plt.close("all")

    def test_vs_plotfft_consistency(self, needs_impl):
        """Both plotfft and plotfft must run on the same signal."""
        from cool_frames.filterbanks import plotfft, plotfft  # type: ignore
        L = 128
        x = np.random.default_rng(7).standard_normal(L)
        F_full = np.fft.fft(x)
        F_real = np.fft.rfft(x)
        try:
            import matplotlib; matplotlib.use("Agg")
            plotfft(F_full)
            plotfft(F_real, real=True, L=L)
        except Exception as exc:
            pytest.fail(f"plotfft / plotfft raised {exc}")
        finally:
            import matplotlib.pyplot as plt
            plt.close("all")

File: tests/layer2_filterbank/unit/TestPlotUtils.m
classdef TestPlotUtils < matlab.unittest.TestCase
%TESTPLOTUTILS  Smoke tests for FFT plotting utilities.
%
%   Covers: plotfft, plotfftreal
%
%   Both functions produce frequency-domain magnitude plots.  These tests
%   verify that they execute without error and accept expected input sizes;
%   they do not validate visual appearance.

    methods (TestClassSetup)
        function setupClass(tc)
            sharedDir = fullfile(fileparts(mfilename('fullpath')), ...
                                 '..', '..', 'shared');
            addpath(sharedDir);
            setup_filterbank_paths();
            % Suppress all figure windows during test run.
            set(0, 'DefaultFigureVisible', 'off');
        end
    end

    methods (TestClassTeardown)
        function teardownClass(~)
            close all;
            set(0, 'DefaultFigureVisible', 'on');
        end
    end

    % ── plotfft ───────────────────────────────────────────────────────────
    methods (Test)

        function testPlotfftRunsWithVector(tc)
            % plotfft must run without error for a full-length FFT vector.
            L = 64;
            f = randn(L, 1);
            F = fft(f);
            tc.verifyWarningFree(@() plotfft(F), ...
                'plotfft: must run without warnings for a length-L FFT vector.');
            close all;
        end

        function testPlotfftRunsWithFs(tc)
            % plotfft must accept an optional sampling-rate argument.
            L = 64;
            fs = 8000;
            F = fft(randn(L, 1));
            try
                plotfft(F, fs);
                close all;
                tc.verifyTrue(true, 'plotfft with fs ran without error.');
            catch ME
                tc.verifyFail(sprintf('plotfft(F,fs) threw: %s', ME.message));
            end
        end

        function testPlotfftRunsWithDynrange(tc)
            % plotfft must accept a dynamic-range limiting argument.
            L = 64;
            F = fft(randn(L, 1));
            try
                plotfft(F, 8000, 60);
                close all;
                tc.verifyTrue(true, 'plotfft with dynrange ran without error.');
            catch ME
                tc.verifyFail(sprintf('plotfft(F,fs,dynrange) threw: %s', ME.message));
            end
        end

        function testPlotfftCreatesAxes(tc)
            % After calling plotfft an axes object should exist.
            L = 128;
            F = fft(randn(L, 1));
            figure('Visible', 'off');
            plotfft(F);
            ax = gca;
            tc.verifyTrue(ishandle(ax) && strcmp(get(ax,'Type'),'axes'), ...
                'plotfft: must produce a valid axes handle.');
            close all;
        end

    end

    % ── plotfftreal ───────────────────────────────────────────────────────
    methods (Test)

        function testPlotfftrealRunsWithVector(tc)
            % plotfftreal must run without error for a positive-frequency vector.
            L = 64;
            f = randn(L, 1);
            F = fftreal(f);         % length floor(L/2)+1
            tc.verifyWarningFree(@() plotfftreal(F), ...
                'plotfftreal: must run without warnings for a fftreal vector.');
            close all;
        end

        function testPlotfftrealRunsWithFs(tc)
            % plotfftreal must accept an optional sampling-rate argument.
            L = 64;
            fs = 8000;
            F = fftreal(randn(L, 1));
            try
                plotfftreal(F, fs);
                close all;
                tc.verifyTrue(true, 'plotfftreal with fs ran without error.');
            catch ME
                tc.verifyFail(sprintf('plotfftreal(F,fs) threw: %s', ME.message));
            end
        end

        function testPlotfftrealRunsWithDynrange(tc)
            % plotfftreal must accept a dynamic-range argument.
            L = 64;
            F = fftreal(randn(L, 1));
            try
                plotfftreal(F, 8000, 60);
                close all;
                tc.verifyTrue(true, 'plotfftreal with dynrange ran without error.');
            catch ME
                tc.verifyFail(sprintf('plotfftreal(F,fs,dynrange) threw: %s', ME.message));
            end
        end

        function testPlotfftrealCreatesAxes(tc)
            % After calling plotfftreal an axes object should exist.
            L = 128;
            F = fftreal(randn(L, 1));
            figure('Visible', 'off');
            plotfftreal(F);
            ax = gca;
            tc.verifyTrue(ishandle(ax) && strcmp(get(ax,'Type'),'axes'), ...
                'plotfftreal: must produce a valid axes handle.');
            close all;
        end

        function testPlotfftrealVsPlotfftConsistency(tc)
            % plotfft and plotfftreal must both accept the same signal and
            % complete without error: no crash for either full or half spectrum.
            L = 128;
            f = randn(L, 1);
            F_full = fft(f);
            F_real = fftreal(f);
            try
                plotfft(F_full);
                close all;
                plotfftreal(F_real);
                close all;
                tc.verifyTrue(true, 'Both plotfft and plotfftreal ran without error.');
            catch ME
                tc.verifyFail(sprintf('Plotting threw: %s', ME.message));
            end
        end

    end

end
